streaming/src/queue/message.cc

```diff
@@ -0,0 +1,240 @@
+#include "message.h"
+
+namespace ray {
+namespace streaming {
+const uint32_t Message::MagicNum = 0xBABA0510;
+
+std::unique_ptr<LocalMemoryBuffer> Message::ToBytes() {
+  uint8_t *bytes = nullptr;
+
+  std::string pboutput;
+  ToProtobuf(&pboutput);
+  int64_t fbs_length = pboutput.length();
+
+  queue::protobuf::StreamingQueueMessageType type = Type();
+  size_t total_len =
+      sizeof(Message::MagicNum) + sizeof(type) + sizeof(fbs_length) + fbs_length;
+  if (buffer_ != nullptr) {
+    total_len += buffer_->Size();
+  }
+  bytes = new uint8_t[total_len];
+  STREAMING_CHECK(bytes != nullptr) << "allocate bytes fail.";
+
+  uint8_t *p_cur = bytes;
+  memcpy(p_cur, &Message::MagicNum, sizeof(Message::MagicNum));
+
+  p_cur += sizeof(Message::MagicNum);
+  memcpy(p_cur, &type, sizeof(type));
+
+  p_cur += sizeof(type);
+  memcpy(p_cur, &fbs_length, sizeof(fbs_length));
+
+  p_cur += sizeof(fbs_length);
+  uint8_t *fbs_bytes = (uint8_t *)pboutput.data();
+  memcpy(p_cur, fbs_bytes, fbs_length);
+  p_cur += fbs_length;
+
+  if (buffer_ != nullptr) {
+    memcpy(p_cur, buffer_->Data(), buffer_->Size());
+  }
+
+  // COPY
+  std::unique_ptr<LocalMemoryBuffer> buffer =
+      std::unique_ptr<LocalMemoryBuffer>(new LocalMemoryBuffer(bytes, total_len, true));
+  delete bytes;
+  return buffer;
+}
+
+void DataMessage::ToProtobuf(std::string *output) {
+  queue::protobuf::StreamingQueueDataMsg msg;
+  msg.set_src_actor_id(actor_id_.Binary());
+  msg.set_dst_actor_id(peer_actor_id_.Binary());
+  msg.set_queue_id(queue_id_.Binary());
+  msg.set_seq_id(seq_id_);
+  msg.set_length(buffer_->Size());
+  msg.set_raw(raw_);
+  msg.SerializeToString(output);
+}
+
+std::shared_ptr<DataMessage> DataMessage::FromBytes(uint8_t *bytes) {
+  bytes += sizeof(uint32_t) + sizeof(queue::protobuf::StreamingQueueMessageType);
+  uint64_t *fbs_length = (uint64_t *)bytes;
+  bytes += sizeof(uint64_t);
+
+  std::string inputpb(reinterpret_cast<char const *>(bytes), *fbs_length);
+  queue::protobuf::StreamingQueueDataMsg message;
+  message.ParseFromString(inputpb);
+  ActorID src_actor_id = ActorID::FromBinary(message.src_actor_id());
+  ActorID dst_actor_id = ActorID::FromBinary(message.dst_actor_id());
+  ObjectID queue_id = ObjectID::FromBinary(message.queue_id());
+  uint64_t seq_id = message.seq_id();
+  uint64_t length = message.length();
+  bool raw = message.raw();
+  bytes += *fbs_length;
+
+  /// COPY
```

Comment:
// Copy data and create a new buffer for streaming queue.

Reply:
> // Copy data and create a new buffer for streaming queue.

fixed.